lib/minstall.js

```diff
@@ -133,315 +118,586 @@ function checkStartConditions() {
     });
-}
 
-function persistExistingConfilctingDependencies(packetsToKeep, moduleFolderName) {
+    // because we check the current path first, and then add the path element, the last element won't be checked.
+    // because of that we just add another pathElement that will be ignored, but will make the last real
+    // pathElement not be ignored
+    possiblePathElements.push('');
+    let currentPath = cwd;
+
+    for (const possiblePathElement of possiblePathElements) {
+      // is this Dependency already on the list of things to install?
+      let installModuleHere = true;
+      for (const modulePath in optimalDependencyTargetFolder) {
+        const modulesToBeInstalled = optimalDependencyTargetFolder[modulePath];
+        const matchingModule = modulesToBeInstalled.find((moduleToBeInstalled) => {
+          return moduleToBeInstalled.identifier === requestedDependency.identifier;
+        });
+
+        if (matchingModule) {
+          logger.debug(`no need to install ${requestedDependency.identifier} to ${currentPath}. a matching version will already be installed to ${modulePath}`);
+          installModuleHere = false;
+          break;
+        }
+      }
+
+      // Is a conflicting dependency-version already installed here?
+      if (installModuleHere) {
+        for (const installedDependency of alreadyInstalledDependencies) {
+          if (installedDependency.name === requestedDependency.name &&
+              installedDependency.location === path.join(currentPath, 'node_modules')) {
+            logger.debug(`${requestedDependency.identifier} can't be installed to ${currentPath}. it conflicts with the already installed ${installedDependency.name}@"${installedDependency.version}"`);
+            installModuleHere = false;
+            break;
+          }
+        }
+      }
+
+      // will any conflicting dependency-versions be installed here?
+      if (installModuleHere && optimalDependencyTargetFolder[currentPath]) {
+        const conflictingDependency = optimalDependencyTargetFolder[currentPath].find((toBeInstalledDependency) => {
+          return toBeInstalledDependency.name === requestedDependency.name &&
+                 toBeInstalledDependency.versionRange !== requestedDependency.versionRange;
+        });
+
+        if (conflictingDependency) {
+          logger.debug(`${requestedDependency.identifier} can't be installed to ${currentPath}. it'd conflict with the to be installed ${conflictingDependency.identifier}`);
+          installModuleHere = false;
+        }
+      }
+
+      if (!installModuleHere) {
+        currentPath = path.join(currentPath, possiblePathElement);
+        continue;
+      }
 
-  const packetFolderNames = packetsToKeep.map((packet) => {
-    return packet.folderName;
+      logger.debug(`found a place to install ${requestedDependency.identifier}: ${currentPath}`);
+
+      // the dependency can be installed here :)
+      if (!optimalDependencyTargetFolder[currentPath]) {
```

Comment:
also a good example, where a semantic variable name would clarify much of the intention here.

Should be refactored; I cannot see what is going on here.

Reply:
You are correct. fyi:
`optimalDependencyTargetFolder[currentPath]` holds the info about all the dependencies that will get installed to the location `currentPath`.  If `currentPath` is not yet a member of the object `optimalDependencyTargetFolder`, it needs to be added before we can push to `optimalDependencyTargetFolder[currentPath]`. This is what this if-block does